### DISCUSSION TITLE
We should able to return complete payload object from the hooks

### DISCUSSION THREAD

**Discussion by kadiresen** [directus/directus#13491] [Archive]
### Describe the Bug

If we modify the payload inside any filter hook, wrong meta values placing over response. For example, current payload is array of 3 items but I've changed it with array as 1 item. Admin app displays data as (and also API's):

![image](https://user-images.githubusercontent.com/6187546/169812643-8906267d-0fae-4abf-844a-67c5ff800ca2.png)

So, pagination operations are not working properly because of that.

### To Reproduce

Create a ```items.read``` filter hook and modify payload object with different number of records.

**Comment by rijkvanzanten** [comment]
The app doesn't rely on the returned `meta` from the endpoints in its pagination (`meta` is no longer in use since we launched the `aggregation` query parameter). 

If you modify the standard output of the API through hooks or other custom logic, the app will start to misbehave in various places (expected and otherwise), so I'd highly recommend setting up a custom endpoint instead if you have a business requirement for custom reading logic 👍🏻

**Reply by kadiresen** [reply]
Thanks for your answer. Sometimes, you wants user to get allowed ones, not all collection items. To do that, I've created a M2O relation with system users against a "tasks" (custom collection), gave a name as "owner". And I've created a ```items.read``` filter hook just runs for "tasks". Logic with that hook is that, If the current user is not Admin and hasn't owns items in payload object, splits them out from payload. But wanted to return ```{ meta: { filter_count: X, total_count: Y , data: some_data } }```, not just ```return data```. Yes, I can do it with a custom endpoint but I can use it just with APIs then.

I'd like to to something like that:

```js
export default ({ filter, action }, { services }) => {
  filter('items.read', async (payload, { query, collection }, { schema, accountability }) => {
    // if current user is not admin and wants to get particular collections that needs ownership
    if (
      accountability &&
      accountability.role !== '84816008-b9cd-47df-ab5b-8aa693ee7c06' &&
      ['tasks'].includes(collection)
    ) {
      // if current user is not the owner of requested items
      if (query.filter?.owner?._eq !== accountability.user) {
        const { ItemsService, MetaService } = services
        const metaService = new MetaService({ accountability, schema })
        const itemsService = new ItemsService(collection, {
          schema,
          accountability,
        })

        // create a new query for items belongs to current user
        const finalQuery = {
          ...query,
          filter: {
            ...query.filter,
            owner: { _eq: accountability.user },
          },
          meta:['filter_count', 'total_count'],
        }

        const meta = await metaService.getMetaForQuery(collection, finalQuery)

        const result = await itemsService.readByQuery(finalQuery)

        return {
          data: result,
          meta: { total_count: meta.filter_count, filter_count: meta.filter_count },
        }
      }
    }
  })
}
```

**Reply by rijkvanzanten** [reply]
> Sometimes, you wants user to get allowed ones, not all collection items.

That's what permissions are for! From your code example, it sounds like you should setup a role with read permissions for `owner` `_eq` `$CURRENT_USER` instead 👍🏻
